meos/src/geo/temporal_rtree.c

```diff
@@ -0,0 +1,862 @@
+  if (! *split)
+  {
+    stbox_expand(new_box, &node->boxes[insertion_node]);
+    *split = false;
+    return;
+  }
+  if (node->count == MAXITEMS)
+  {
+    *split = true;
+    return;
+  }
+  RTreeNode *right;
+  node_split(rtree, node->nodes[insertion_node], &node->boxes[insertion_node],
+    &right);
+  node_box_calculate(rtree, node->nodes[insertion_node], &node->boxes[insertion_node]);
+  node_box_calculate(rtree, right, &node->boxes[node->count]);
+  node->nodes[node->count] = right;
+  node->count++;
+  node_insert(rtree, node_bounding_box, node, new_box, id, split);
+  return;
+}
+
+/**
+ * @brief Checks if a number bigger than 0 is a power of two or not. 
+ * @param[in] n Number to check
+ * @returns True if `n` is a power of two, False otherwise.
+ */
+static inline bool
+is_power_of_two(const int n)
+{
+  return (n & (n - 1)) == 0;
+}
+
+/**
+ * @brief Adds an ID to the dynamically allocated array with the answer of a 
+ * query.
+ * @param[in] id The integer ID to be added to the array.
+ * @param[in] ids Pointer to a pointer to the dynamically allocated array of
+ * integers.
+ * @param[in] count Pointer to an integer representing the current number of
+ * elements in the array.
+ */
+static void
+add_answer(const int id, int **ids, int *count)
+{
+  /* Every power of two that exceeds the size of the array must  
+   * be resized to double the current size */
+  if (*count >= SEARCH_ARRAY_STARTING_SIZE && is_power_of_two(*count))
+    *ids = repalloc(*ids, sizeof(int) * (*count) * 2);
+  (*ids)[*count] = id;
+  (*count)++;
+  return;
+}
+
+/**
+ * @brief Frees the memory allocated for an RTree node and its associated
+ * resources.
+ * @details This function recursively frees the memory of an RTree node.
+ * If the node is a branch node, it first recursively frees all child nodes. 
+ * After handling the child nodes, it frees the memory allocated for the
+ * STBoxes and the arrays of boxes and child nodes within the current node.
+ * Finally, it frees the memory allocated for the node itself.
+ * @param[in] node Pointer to the RTreeNode structure to be freed.
+ */
+static void
+node_free(RTreeNode *node)
+{
+  if (node->isRoot == RTREE_INNER_NODE)
+  {
+    for (int i = 0; i < node->count; ++i)
+      node_free(node->nodes[i]);
+  }
+  free(node);
+  return;
+}
+
+/**
+ * @brief Searches recursively a node looking for hits with a query.
+ * @param[in] node The node to be searched
+ * @param[in] query The STBox that serves as query
+ * @param[in] ids The array with the list of answers.
+ * @param[in] count Total of elements found
+ */
+void node_search(const RTreeNode *node, const STBox *query, int **ids,
+  int *count)
+{
+  for (int i = 0; i < node->count; ++i)
+  {
+    if (overlaps_stbox_stbox(query, &node->boxes[i]))
+    {
+      if (node->isRoot == RTREE_ROOT_NODE)
+        add_answer(node->ids[i], ids, count);
+      else
+        node_search(node->nodes[i], query, ids, count);
+    }
```

Comment:
This doesn't look right. the id's should be at the leaves, not at the root.

Reply:
This was already on the master branch. 
https://github.com/MobilityDB/MobilityDB/blob/master/meos/src/geo/tspatial_rtree.c#L574-L587
Could you please specify the problem ?

Reply:
Well, an internal node stores pointers to its children (`node->nodes`, which maybe should be renamed for clarity). The id's are stored in the leaf nodes (`node->ids`). Thus, we need something like:

```
if (node->nodetype == RTREE_LEAF_NODE)
    add_answer(node->ids[i], ids, count);
else
    node_search(node->nodes[i], query, ids, count);
```

Here, we always directly return the ids from the root node, which is incorrect.

Reply:
Where in the algorithm the nodetype should be set to `RTREE_LEAF_NODE` to be used here ? See also comment below.